geostore/s3_utils.py

```diff
@@ -52,3 +54,58 @@ def s3_url_reader(staging_url: str) -> GeostoreS3Response:
     staging_s3_client = get_s3_client_for_role(s3_role_arn)
     geostore_s3_client = get_s3_client_for_role(get_param(ParameterName.S3_USERS_ROLE_ARN))
     return s3_url_reader
+
+
+def check_if_s3_object_exists(bucket: str, key: str) -> bool:
+    s3_resource = boto3.resource("s3")
+
+    try:
+        s3_resource.Object(bucket, key).load()
+    except ClientError:
+        # 404 is most likely the response code here
+        # other response code is possible, but we return false rather than raise,
+        # thus allowing the next step to continue rather than stalling the entire process
+        return False
+
+    return True
+
+
+def get_s3_etag(s3_bucket: str, s3_object_key: str, logger: Logger) -> str:
+    s3_client = boto3.client("s3")
+
+    try:
+        s3_response = s3_client.head_object(Bucket=s3_bucket, Key=s3_object_key)
+        s3_etag = s3_response["ETag"].strip('"')
+        return s3_etag
+    except ClientError as error:
+        logger.error(
+            f"Unable to fetch eTag for {s3_object_key} in s3://{s3_bucket},"
+            f"due to {error}, even though object exists in bucket.",
+            extra={GIT_COMMIT: get_param(ParameterName.GIT_COMMIT)},
+        )
+        # rather than raise, we return an empty string, indicating that the etag is different
+        # thus allowing the next step to continue rather than stalling the entire process
+        return ""
+
+
+def calculate_s3_etag(
+    body: bytes,
+    # https://docs.aws.amazon.com/cli/latest/topic/s3-config.html#multipart-chunksize
+    chunk_size: int = 8 * 1024 * 1024,  # Default value
```

Comment:
>One way to make sure we're actually implementing an ETag calculation which matches AWS would be to have tests which upload empty, single-chunk and multi-chunk files to AWS...

This thought did occurred to me as I was working on this. I wasn't too sure about having (non trivial size) tests files either stored in a bucket somewhere or added to the git repo. Generating random file content is much more appealing, but I was under the impression that `md5` is used as part of calculating the eTag? Perhaps I have misunderstood something here??

I ended up writing the test using a byte representation of files (see `test_s3_utils.py::should_return_accurate_etag_value`), which appears to be enough for me to calculate the corresponding `etag` for testing purposes. Would you recommend I revisit this with the suggestions above? (i.e. either import asset file for testing or generate random file content)

>Also, should this really be configurable?

I saw a reference somewhere that suggests different clients use different default values (e.g. using aws cli vs third party software). Unfortunately I have lost the link that suggest this and a quick search comes up blank. That said, I am aware that this is unlikely to affect _geostore_ since s3 to s3 transfer is the only currently supported transfer method. I have left this configurable with default value mostly to allow the test to simulate a multi chunk file without having to upload or generate files larger than 8mb.

Perhaps a discussion might be helpful here?

Reply:
> > One way to make sure we're actually implementing an ETag calculation which matches AWS would be to have tests which upload empty, single-chunk and multi-chunk files to AWS...
> 
> This thought did occurred to me as I was working on this. I wasn't too sure about having (non trivial size) tests files either stored in a bucket somewhere or added to the git repo. Generating random file content is much more appealing, but I was under the impression that `md5` is used as part of calculating the eTag? Perhaps I have misunderstood something here??

We wouldn't create any files in the filesystem - simply push a stream of `os.urandom` to an S3 bucket as part of the test. We don't really care what the contents are (except for the empty file test), just that it uses our upload functionality and that the ETag we calculate is the same as reported by AWS.

> I ended up writing the test using a byte representation of files (see `test_s3_utils.py::should_return_accurate_etag_value`), which appears to be enough for me to calculate the corresponding `etag` for testing purposes. Would you recommend I revisit this with the suggestions above? (i.e. either import asset file for testing or generate random file content)

Yes. Basically, there's no easy way for someone reading the tests to say "yes, this test does what it says it should do", because the inputs and expected outputs have no connection to each other which can be easily understood. If we instead have a test which just sends some random data to S3 and then just asserts that our ETag calculation gives the same result as S3, then we've proven just that: our implementation is (at least for the values we're testing) equivalent to the AWS implementation. Which I think is the best we can hope for, since if we ever start using a different chunk size or some other way to upload things, we'll end up with a different ETag calculation.

(Ideally we'd simply call a library function which implements ETag calculation, but I guess you've not been able to find a good/official one? I could only find third party ones. :shrug:)

> > Also, should this really be configurable?
> 
> I saw a reference somewhere that suggests different clients use different default values (e.g. using aws cli vs third party software). Unfortunately I have lost the link that suggest this and a quick search comes up blank. That said, I am aware that this is unlikely to affect _geostore_ since s3 to s3 transfer is the only currently supported transfer method. I have left this configurable with default value mostly to allow the test to simulate a multi chunk file without having to upload or generate files larger than 8mb.

I don't think such a test is very useful, because it's not testing the core of the algorithm - that we get the same result from our implementation as from S3 itself.